swarm_dial.go

```diff
@@ -281,39 +286,306 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
 	return nil, err
 }
 
-// doDial is an ugly shim method to retain all the logging and backoff logic
-// of the old dialsync code
-func (s *Swarm) doDial(ctx context.Context, p peer.ID) (*Conn, error) {
-	// Short circuit.
-	// By the time we take the dial lock, we may already *have* a connection
-	// to the peer.
-	c := s.bestAcceptableConnToPeer(ctx, p)
-	if c != nil {
-		return c, nil
+///////////////////////////////////////////////////////////////////////////////////
+// lo and behold, The Dialer
+// TODO explain how all this works
+//////////////////////////////////////////////////////////////////////////////////
+type DialRequest struct {
+	Ctx   context.Context
+	Resch chan DialResponse
+}
+
+type DialResponse struct {
+	Conn *Conn
+	Err  error
+}
```

Comment:
Unfortunately the tests are in a different package and `dial_sync_test.go` needs access to them.
They weren't public initially, I turned them public for the tests.

Reply:
Yeah, I don't like having them public either.

In principle it should be simple, only `dial_sync_test` needs them and that doesn't do any cross-package imports, so it can be moved into the `swarm` package itself.

However, all the other tests need to be in a separate package and we won't be able to mix the two test packages together in the same directory.
So we might have to move all the other tests into a separate directory.

Reply:
Ah we don't have to move them, `go test` apparently can mix and match different packages.

Reply:
done, I made them private and moved `dial_sync_test` to the `swarm` package.